### PR TITLE
Light SDK

### PR DESCRIPTION
**Pull request by adrianbrink**
## Describe your changes

Implements a first version of a simpler, more user-friendly SDK.

## Indicate on which release or other PRs this topic is based on

Commit 74d412e14 of `base` (v0.28.1)

## Checklist before merging to `draft`
- [x] I have added a changelog
- [x] Git history is in acceptable state

### CODE REVIEW

File: light_sdk/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+    "tx_update_steward_commission.wasm";
+/// Redelegate transaction WASM path
+pub const TX_REDELEGATE_WASM: &str = "tx_redelegate.wasm";
+/// Target chain ID
+pub const CHAIN_ID: &str = "localnet.3e837878d84b54a40f-0";
+/// Reveal public key transaction code hash
+pub const TX_REVEAL_PK_HASH: [u8; 32] = [0x8d, 0x4e, 0x20, 0x08, 0x30, 0xa9, 0xf6, 0x03, 0xd6, 0x47, 0x79, 0x9f, 0xaf, 0xdb, 0xf1, 0x94, 0x6e, 0x07, 0xa0, 0x38, 0xa6, 0x7c, 0xfa, 0x09, 0xd4, 0xee, 0xb3, 0x1c, 0x80, 0xc7, 0x74, 0xce];
+
+/// Takes any kind of inner tx and gives me back my sign bytes
+pub fn get_inner_sign_bytes(mut tx: Tx) -> Hash {
```
Comment (adrianbrink) [inline light_sdk/src/lib.rs:155]: These should be immutable functions on a specific RevealPk struct. Ie. this signature should be repeated for every type of transaction.

Also this function should be immutable.

File: light_sdk/src/tx_builders/mod.rs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+use namada_core::types::token;
+use namada_core::types::token::{Amount, DenominatedAmount, MaspDenom};
+use namada_core::types::transaction::Fee;
+use namada_core::types::transaction::GasLimit;
+use std::collections::BTreeMap;
+use std::str::FromStr;
+
+pub mod pos;
+
+pub(in crate::tx_builders) fn build_tx(
```
Comment (adrianbrink) [inline light_sdk/src/tx_builders/mod.rs:25]: no generics - everything should be passed in as strings

File: light_sdk/src/tx_builders/mod.rs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        secret_keys,
+        public_keys_index_map,
+        signer,
+    );
+
+    (tx, signatures)
+}
+
+pub(in crate::tx_builders) fn attach_raw_signatures(
+    mut tx: Tx,
```
Comment (adrianbrink) [inline light_sdk/src/tx_builders/mod.rs:60]: no mutability - we should consume the tx object and return a new one

File: light_sdk/src/tx_builders/pos.rs [hunk 4]
```diff
@@ -0,0 +1,653 @@
...
+            code_hash,
+            TX_UNBOND_WASM.to_string(),
+            chain_id,
+        ))
+    }
+
+    /// Generate the signature(s) for the given transaction and signers
+    pub fn generate_signatures(
+        mut self,
+        secret_keys: &[common::SecretKey],
```
Comment (adrianbrink) [inline light_sdk/src/tx_builders/pos.rs:139]: the library doesn't take secret keys. 

it should expose `sign_bytes()` and then the caller figures out how to sign it

File: light_sdk/src/tx_builders/pos.rs [hunk 5]
```diff
@@ -0,0 +1,653 @@
...
+            timestamp,
+            expiration,
+            code_hash,
+            TX_INIT_VALIDATOR_WASM.to_string(),
+            chain_id,
+        ))
+    }
+
+    /// Generate the signature(s) for the given transaction and signers
+    pub fn generate_signatures(
```
Comment (adrianbrink) [inline light_sdk/src/tx_builders/pos.rs:213]: likewise - this kind of function shouldn't exist in the public api

File: light_sdk/src/tx_builders/mod.rs (hunk 3 above)
Comment (grarco) [reply light_sdk/src/tx_builders/mod.rs:60]: That's what I'm doing here but I need the object to be mutable in the scope of the function

File: light_sdk/src/tx_builders/mod.rs (hunk 2 above)
Comment (grarco) [reply light_sdk/src/tx_builders/mod.rs:25]: This function is not exposed to the SDK user

File: light_sdk/src/reading/mod.rs [hunk 6]
```diff
@@ -0,0 +1,679 @@
...
+            })?
+            .data;
+        println!("Dry-run result: {}", result);
+        Ok(result)
+    }
+
+    /// Lookup the full response accompanying the specified transaction event
+    pub fn query_tx_response(
+        tendermint_addr: &str,
+        tx_query: TxEventQuery<'_>,
```
Comment (adrianbrink) [inline light_sdk/src/reading/mod.rs:99]: This is essentially a hash, right? Can this type be instantiated by the caller?

File: light_sdk/src/reading/mod.rs [hunk 7]
```diff
@@ -0,0 +1,679 @@
...
+            TendermintAddress::from_str(tendermint_addr)
+                .map_err(|e| Error::Other(e.to_string()))?,
+        )
+        .map_err(|e| Error::Other(e.to_string()))?;
+        let rt = Runtime::new().unwrap();
+        rt.block_on(rpc::query_bond(&client, source, validator, epoch))
+    }
+
+    /// Query a validator's unbonds for a given epoch
+    pub fn query_and_print_unbonds(
```
Comment (adrianbrink) [inline light_sdk/src/reading/mod.rs:446]: let's not have print options in the api. if a caller wants to print something, they can do it in their own code.

File: light_sdk/src/reading/mod.rs [hunk 8]
```diff
@@ -0,0 +1,679 @@
+use std::str::FromStr;
```
Comment (adrianbrink) [inline light_sdk/src/reading/mod.rs:1]: Overall I like the api.

It's probably better to split the individual submodules into their own files.

File: light_sdk/src/writing/mod.rs [hunk 9]
```diff
@@ -0,0 +1,3 @@
+pub mod blocking {}
```
Comment (adrianbrink) [inline light_sdk/src/writing/mod.rs:1]: This should be just 1 function - submit_tx().

File: light_sdk/src/reading/mod.rs (hunk 6 above)
Comment (batconjurer) [reply light_sdk/src/reading/mod.rs]: yeah, I changed this to take a `&str`. It means you can only query applied txs and not accepted, but I think that is totally fine.

**Review by batconjurer — DISMISSED** [review]

**Pull request merged**